test/integration/mpi_job_controller_test.go

```diff
@@ -175,6 +175,132 @@ func TestMPIJobSuccess(t *testing.T) {
 	}
 }
 
+func TestMPIJobWaitWorkers(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+	s := newTestSetup(ctx, t)
+	startController(ctx, s.kClient, s.mpiClient, nil)
+
+	mpiJob := &kubeflow.MPIJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "job",
+			Namespace: s.namespace,
+		},
+		Spec: kubeflow.MPIJobSpec{
+			SlotsPerWorker: newInt32(1),
+			WaitForWorkers: true,
+			RunPolicy: kubeflow.RunPolicy{
+				CleanPodPolicy: kubeflow.NewCleanPodPolicy(kubeflow.CleanPodPolicyRunning),
+			},
+			MPIReplicaSpecs: map[kubeflow.MPIReplicaType]*common.ReplicaSpec{
+				kubeflow.MPIReplicaTypeLauncher: {
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name:  "main",
+									Image: "mpi-image",
+								},
+							},
+						},
+					},
+				},
+				kubeflow.MPIReplicaTypeWorker: {
+					Replicas: newInt32(2),
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name:  "main",
+									Image: "mpi-image",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	var err error
+	mpiJob, err = s.mpiClient.KubeflowV2beta1().MPIJobs(s.namespace).Create(ctx, mpiJob, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed sending job to apiserver: %v", err)
+	}
+
+	s.events.expect(eventForJob(corev1.Event{
+		Type:   corev1.EventTypeNormal,
+		Reason: "MPIJobCreated",
+	}, mpiJob))
+
+	mpiJob = validateMPIJobStatus(ctx, t, s.mpiClient, mpiJob, map[kubeflow.MPIReplicaType]*kubeflow.ReplicaStatus{
+		kubeflow.MPIReplicaTypeWorker:   {},
+	})
+	if !mpiJobHasCondition(mpiJob, kubeflow.JobCreated) {
+		t.Errorf("MPIJob missing Created condition")
+	}
+	s.events.verify(t)
+
+	workerPods, err := getPodsForJob(ctx, s.kClient, mpiJob)
+	if err != nil {
+		t.Fatalf("Cannot get worker pods from job: %v", err)
+	}
+
+	err = updatePodsToPhase(ctx, s.kClient, workerPods, corev1.PodRunning)
+	if err != nil {
+		t.Fatalf("Updating worker Pods to Running phase: %v", err)
+	}
+
+	// No launcher here, workers are running, but not ready yet
+        validateMPIJobStatus(ctx, t, s.mpiClient, mpiJob, map[kubeflow.MPIReplicaType]*kubeflow.ReplicaStatus{
+		kubeflow.MPIReplicaTypeWorker: {
+			Active: 2,
+		},
+	})
+	
```

Comment:
One more nit.
Can we check whether the launcher Job isn't created?

Reply:
I am no expert here, but from debugs it seems that this function passes if and only if exact set of object is created. I.e., there must be only 2 active workers and nothing else, or it fails in diff such as there is launcher in advance. Isn't it so?

Reply:
IIUC, if we try to get the launcher Job here, we will receive a NotFound error.
And then, the behavior is correct.

Reply:
I meant we don't want to check the launcher Job using validateMPIJobStatus.

Reply:
Yes, correct.
My concern is the case of there are some bugs in MPIJob's status update process.

Reply:
So, I would suggest checking the existence of the launcher Job in the following:

```golang
_, err := s.kClient.BatchV1().Jobs(launcherPod.Namespace).Get(s.namespace, "job-launcher")
if !apierrors,IsNotfound() {
  t.Fatalf(ERROR STATEMENT)
}
```

Reply:
ok, fixed